sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryMetrics.java

```diff
@@ -0,0 +1,907 @@
+                if (latency != null) {
+                    Timer requestLatencyMeter = Timer
+                        .builder(nameOf("req.rntbd.latency"))
+                        .description("RNTBD Request latency")
+                        .maximumExpectedValue(Duration.ofSeconds(6))
+                        .publishPercentiles(0.95, 0.99)
+                        .publishPercentileHistogram(true)
+                        .tags(requestTags)
+                        .register(compositeRegistry);
+                    requestLatencyMeter.record(latency);
+                }
+
+                recordRequestTimeline(
+                    "req.rntbd.timeline.",
+                    storeResponseDiagnostics.getRequestTimeline(), requestTags);
+
+                // TODO @fabianm - delete completely if we agree in PR that this metric isn't needed
+                /*
+                recordRntbdChannelAcquisitionTimeline(
+                    storeResponseDiagnostics.getChannelAcquisitionTimeline(),
+                    requestTags);*/
+
+                recordRequestPayloadSizes(
+                    storeResponseDiagnostics.getRequestPayloadLength(),
+                    storeResponseDiagnostics.getResponsePayloadLength()
+                );
+
+                // TODO @fabianm - delete completely if we agree in PR that this metric isn't needed
+                /*
+                recordRntbdChannelStatistics(
+                    storeResponseDiagnostics.getPendingRequestQueueSize(),
+                    storeResponseDiagnostics.getRntbdChannelTaskQueueSize(),
+                    requestTags
+                );*/
+
+                recordRntbdEndpointStatistics(
+                    storeResponseDiagnostics.getRntbdEndpointStatistics(),
+                    requestTags);
+            }
+        }
+
+        private void recordGatewayStatistics(
+            Duration latency,
+            ClientSideRequestStatistics.GatewayStatistics gatewayStatistics) {
+
+            if (gatewayStatistics == null) {
+                return;
+            }
+
+            EnumSet<TagName> metricTagNamesForGateway = metricTagNames.clone();
+            metricTagNamesForGateway.remove(TagName.RegionName);
+            metricTagNamesForGateway.remove(TagName.ServiceAddress);
+            metricTagNamesForGateway.remove(TagName.ServiceEndpoint);
+
+            Tags requestTags = operationTags.and(
+                createRequestTags(
+                    metricTagNamesForGateway,
+                    gatewayStatistics.getPartitionKeyRangeId(),
+                    gatewayStatistics.getStatusCode(),
+                    gatewayStatistics.getSubStatusCode(),
+                    gatewayStatistics.getResourceType(),
+                    gatewayStatistics.getOperationType(),
+                    null,
+                    null,
+                    null)
+            );
+
+            if (latency != null) {
+                Timer requestLatencyMeter = Timer
```

Comment:
We are creating timer on every request , is there we these can be saved and accessed from some sort of pool ?

Reply:
No - the Micrometer API is built so that you getOrCreate timers, gauges etc. depending on the tags and config you provide. It is using a cache behind the scenes and using the config as a cache key. So this is the expected calling pattern for their API.